metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/SyncService.scala

```diff
@@ -158,49 +169,177 @@ class SyncService[F[_]: Sync, N, A <: Agreement](
+            case Some(isValid) =>
+              syncPipe.send(SyncPipe.PrepareResponse(request, isValid))
+            case None =>
+              // We didn't have data to decide validity in time; not responding.
+              ().pure[F]
+          }
+        }
+        .void
   }
 
-  // TODO (PM-3132, PM-3133): Block validation.
-  private def validateBlock(block: A#Block): F[Boolean] = ???
+  /** Shut down the any outstanding block downloads, sync the view,
+    * then create another block synchronizer instance to resume with.
+    */
+  private def handleStatusRequest(
+      makeBlockSync: F[BlockSync],
+      blockSync: BlockSync,
+      viewSynchronizer: ViewSynchronizer[F, A],
+      request: SyncPipe.StatusRequest
+  ): F[(BlockSync, ViewNumber)] =
+    for {
+      // Cancel all outstanding block syncing.
+      _ <- blockSync.fiberMapRelease
+      // The block synchronizer is still usable.
+      viewNumber <- syncStatus(
+        blockSync.synchronizer,
+        viewSynchronizer
+      ).handleErrorWith { case NonFatal(ex) =>
+        tracers.error(ex).as(request.viewNumber)
+      }
+      // Create a fresh fiber and block synchronizer instance.
+      // When the previous goes out of scope, its ephemeral storage is freed.
+      newBlockSync <- makeBlockSync
+    } yield (newBlockSync, viewNumber)
+
+  /** Get the latest status of federation members, download the corresponding block
+    * and prune all existing block history, making the latest Commit Q.C. the new
+    * root in the block tree.
+    *
+    * This is done in the foreground, no further requests are taken from the pipe.
+    */
+  private def syncStatus(
+      blockSynchronizer: BlockSynchronizer[F, N, A],
+      viewSynchronizer: ViewSynchronizer[F, A]
+  ): F[ViewNumber] =
+    for {
+      // Sync to the latest Commit Q.C.
+      federationStatus <- viewSynchronizer.sync
+      status = federationStatus.status
+
+      // Download the block in the Commit Q.C.
+      block <- blockSynchronizer
+        .getBlockFromQuorumCertificate(
+          federationStatus.sources,
+          status.commitQC
+        )
+        .rethrow
+
+      // Sync any application specific state, e.g. a ledger.
+      // Do this before we prune the existing blocks and set the new root.
+      _ <- appService.syncState(federationStatus.sources, block)
+
+      // Prune the block store from earlier blocks that are no longer traversable.
+      _ <- fastForwardStorage(status, block)
+
+      // Tell the ConsensusService about the new Status.
+      _ <- syncPipe.send(SyncPipe.StatusResponse(status))
+    } yield status.viewNumber
+
+  /** Replace the state we have persisted with what we synced with the federation.
+    *
+    * Prunes old blocks, the Commit Q.C. will be the new root.
+    */
+  private def fastForwardStorage(status: Status[A], block: A#Block): F[Unit] = {
+    val blockHash = Block[A].blockHash(block)
+    assert(blockHash == status.commitQC.blockHash)
```

Comment:
We'll have a lot of those. Do you think it would've been better if we had a separate QC type for each phase?
Just curious about your opinion, it's certainly not the right time to change it.

Reply:
Not sure; I started with having different message types for them, but [merged them](https://github.com/input-output-hk/metronome/pull/5/commits/0004fbd40246736e6b8cf520473ddad22aa49e23) because they looked so similar.  

Maybe we could have some phantom type for it even now, something like:
```scala
type Commit = Commit.type 
...
case class QuorumCertificate[A <: Agreement, P <: VotingPhase](
    phase: P,
    viewNumber: ViewNumber,
    blockHash: A#Hash,
    signature: GroupSignature[A#PKey, (VotingPhase, ViewNumber, A#Hash), A#GSig]
)
```
The codecs can make sure that things like `Status` get the right kind then.
Not sure how we can coerce them when they appear in arbitrary places like the `Quorum` message though.

Reply:
Tried; it's okay in the code but gets annoying in the `Gen` instances where there's a lot of `phase.next` and similar stepping.

Reply:
Maybe it could be a tagged type, so we can use it in some places, ignore it in others.

Reply:
Nah, that doesn't work either :-/

Reply:
Made it work, at least up to the PR with the block executor in https://github.com/input-output-hk/metronome/pull/46